### PR TITLE
[Fuser] Do not attempt to use OpenMP if build without OpenMP support

### PR DESCRIPTION
**Pull request by malfet**
Clang from XCode does not support `-fopenmp` option, no need to try to compile with it.
Infer whether OpenMP is supported by checking _OPENMP define.
Also, use clang compiler if host app was compiled with clang rather than gcc.
Fix few range loop warnings and add static_asserts that range loop variables are raw pointers.

This changes makes fuser tests on OS X a bit faster.

Before:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... clang: error: unsupported option '-fopenmp'
clang: error: unsupported option '-fopenmp'
warning: pytorch jit fuser failed to compile with openmp, trying without it...
ok

----------------------------------------------------------------------
Ran 1 test in 0.468s

OK
```

After:
```
% python3 test_jit.py -v  TestScript.test_batchnorm_fuser_cpu
Fail to import hypothesis in common_utils, tests are not derandomized
CUDA not available, skipping tests
test_batchnorm_fuser_cpu (__main__.TestScript) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.435s

OK
```

### CODE REVIEW

**Review by janeyx99 — APPROVED** [review]
interesting, looks fine

File: torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp [hunk 1]
```diff
@@ -196,11 +196,18 @@ struct CompilerConfig {
...
   const std::string openmp_flags = "/openmp";
+#elif defined(__APPLE__)
+  std::string cxx = "clang++";
+  const std::string openmp_flags = "";
 #else
   std::string cxx = "g++";
   const std::string openmp_flags = "-fopenmp";
 #endif
+#if defined(__APPLE__)
+  bool openmp = false;
```
Comment (smessmer) [inline torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp:207]: can we add a comment for posterity on why we're doing this?

**Review by smessmer — APPROVED** [review]

File: torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp (hunk 1 above)
Comment (malfet) [reply torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp:207]: Sure, OpenMP support is MacOS devtools is explicitly disabled:
```
   $ clang -c omp.c -fopenmp
   clang: error: unsupported option '-fopenmp'
```

**Review by walterddr — APPROVED** [review]

File: torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp (hunk 1 above)
Comment (malfet) [reply torch/csrc/jit/codegen/fuser/cpu/fused_kernel.cpp]: Updated the PR to rely less on the platform, but more on the compilation flags (i.e. openmp is set to false if _OPENMP flag is not defined)

**Pull request closed without merging**